File: frontend/logging/src/lib/utils/bearing.test.ts
import { describe, it, expect } from 'vitest';
import {
    calculateBearing,
    gridToDecimal,
    haversineKm,
    pathInfo,
} from './bearing';

describe('gridToDecimal', () => {
    it('returns null for empty input', () => {
        expect(gridToDecimal('')).toBeNull();
    });

    it('returns null for invalid grid', () => {
        expect(gridToDecimal('ZZ99')).toBeNull();
    });

    it('case-folds input', () => {
        const a = gridToDecimal('JN58td');
        const b = gridToDecimal('JN58TD');
        expect(a).not.toBeNull();
        expect(b).not.toBeNull();
        expect(a!.lat).toBeCloseTo(b!.lat, 9);
        expect(a!.lon).toBeCloseTo(b!.lon, 9);
    });

    // v1 reference vectors from station-manager-v1/internal/maidenhead/bearing_test.go
    it('matches v1 vectors', () => {
        const munich = gridToDecimal('JN58td');
        expect(munich).not.toBeNull();
        expect(munich!.lat).toBeCloseTo(48.1458333, 4);
        expect(munich!.lon).toBeCloseTo(11.625, 4);

        const aa00aa = gridToDecimal('AA00aa');
        expect(aa00aa).not.toBeNull();
        expect(aa00aa!.lat).toBeCloseTo(-89.9791667, 4);
        expect(aa00aa!.lon).toBeCloseTo(-179.9583333, 4);

        const jj00aa = gridToDecimal('JJ00aa');
        expect(jj00aa).not.toBeNull();
        expect(jj00aa!.lat).toBeCloseTo(0.0208333, 4);
        expect(jj00aa!.lon).toBeCloseTo(0.0416667, 4);
    });

    it('handles 4-char input', () => {
        // IO91 corner is (-2, 51); centre is (-1, 51.5).
        const io91 = gridToDecimal('IO91');
        expect(io91).not.toBeNull();
        expect(io91!.lat).toBeCloseTo(51.5, 9);
        expect(io91!.lon).toBeCloseTo(-1.0, 9);
    });

    it('handles 8-char input', () => {
        const io91vl42 = gridToDecimal('IO91vl42');
        expect(io91vl42).not.toBeNull();
        // 6-char IO91vl centre is well-defined; 8-char shifts within
        // the subsquare. Just check we got finite numbers in range.
        expect(io91vl42!.lat).toBeGreaterThan(-90);
        expect(io91vl42!.lat).toBeLessThan(90);
        expect(io91vl42!.lon).toBeGreaterThan(-180);
        expect(io91vl42!.lon).toBeLessThan(180);
    });
});

describe('calculateBearing', () => {
    it('returns 0 for due north', () => {
        const b = calculateBearing(0, 0, 10, 0);
        expect(b).toBeGreaterThanOrEqual(-1);
        expect(b).toBeLessThanOrEqual(1);
    });

    it('returns 90 for due east', () => {
        const b = calculateBearing(0, 0, 0, 10);
        expect(b).toBeGreaterThanOrEqual(89);
        expect(b).toBeLessThanOrEqual(91);
    });

    it('returns 180 for due south', () => {
        const b = calculateBearing(10, 0, 0, 0);
        expect(b).toBeGreaterThanOrEqual(179);
        expect(b).toBeLessThanOrEqual(181);
    });

    it('returns 270 for due west', () => {
        const b = calculateBearing(0, 10, 0, 0);
        expect(b).toBeGreaterThanOrEqual(269);
        expect(b).toBeLessThanOrEqual(271);
    });

    // London → New York, ~288.3° per v1's spherical model.
    it('matches v1 London→NY reference (~288.3°)', () => {
        const b = calculateBearing(51.5074, -0.1278, 40.7128, -74.006);
        expect(Math.abs(b - 288.3)).toBeLessThanOrEqual(1.0);
    });

    it('rounds to 0.1°', () => {
        // Result has at most one digit past the decimal point.
        const b = calculateBearing(51.5, -1, 48.15, 11.625);
        expect(b * 10).toBeCloseTo(Math.round(b * 10), 9);
    });
});

describe('haversineKm', () => {
    it('returns 0 for identical points', () => {
        expect(haversineKm(51.5, -1, 51.5, -1)).toBe(0);
    });

    it('matches a known great-circle distance (London→Paris ~344 km)', () => {
        const d = haversineKm(51.5074, -0.1278, 48.8566, 2.3522);
        // Math.ceil rounding; allow ±2 km for endpoint precision.
        expect(d).toBeGreaterThanOrEqual(342);
        expect(d).toBeLessThanOrEqual(346);
    });
});

describe('pathInfo', () => {
    it('returns null for invalid local grid', () => {
        expect(pathInfo('ZZ99', 'JN58td')).toBeNull();
    });

    it('returns null for invalid remote grid', () => {
        expect(pathInfo('JN58td', 'ZZ99')).toBeNull();
    });

    it('returns null for empty input', () => {
        expect(pathInfo('', 'JN58td')).toBeNull();
        expect(pathInfo('JN58td', '')).toBeNull();
    });

    it('short + long bearings are 180° apart', () => {
        const p = pathInfo('JN58td', 'FN31pr');
        expect(p).not.toBeNull();
        // Either spb + 180 ≡ lpb (mod 360) or lpb + 180 ≡ spb (mod 360).
        const a = ((p!.shortPathBearing + 180) % 360);
        const b = ((p!.longPathBearing + 180) % 360);
        const diffA = Math.abs(a - p!.longPathBearing);
        const diffB = Math.abs(b - p!.shortPathBearing);
        expect(Math.min(diffA, diffB)).toBeLessThanOrEqual(0.2);
    });

    it('short + long km approx Earth circumference', () => {
        const p = pathInfo('JN58td', 'FN31pr');
        expect(p).not.toBeNull();
        const sum = p!.shortPathDistanceKm + p!.longPathDistanceKm;
        const earthCircumference = Math.ceil(2 * Math.PI * 6371);
        // v1 tolerance was 2 km; ceil-rounding can drift a bit more
        // when both halves are independently ceiled — allow 3 km.
        expect(Math.abs(sum - earthCircumference)).toBeLessThanOrEqual(3);
    });

    it('miles match km via 0.621371 ceil-rounded', () => {
        const p = pathInfo('JN58td', 'FN31pr');
        expect(p).not.toBeNull();
        expect(p!.shortPathDistanceMiles).toBe(
            Math.ceil(p!.shortPathDistanceKm * 0.621371),
        );
        expect(p!.longPathDistanceMiles).toBe(
            Math.ceil(p!.longPathDistanceKm * 0.621371),
        );
    });

    it('produces sensible bearings for JN58td→FN31pr (Munich→New England)', () => {
        const p = pathInfo('JN58td', 'FN31pr');
        expect(p).not.toBeNull();
        // Roughly westward (NY is west of Munich).
        expect(p!.shortPathBearing).toBeGreaterThan(180);
        expect(p!.shortPathBearing).toBeLessThan(360);
    });
});

File: internal/utils/bearing_test.go
package utils

import (
	"math"
	"testing"
)

// These vectors mirror the SPA's frontend/logging/src/lib/utils/bearing.test.ts
// so the daemon-side QSO geometry (FT8 ANT_AZ / DISTANCE / ANT_PATH) matches what
// the SPA computes for the same grids.

func TestInitialBearing_Cardinals(t *testing.T) {
	cases := []struct {
		name                   string
		lat1, lon1, lat2, lon2 float64
		want                   float64
	}{
		{"north", 0, 0, 10, 0, 0},
		{"east", 0, 0, 0, 10, 90},
		{"south", 10, 0, 0, 0, 180},
		{"west", 0, 10, 0, 0, 270},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := initialBearing(c.lat1, c.lon1, c.lat2, c.lon2)
			if math.Abs(got-c.want) > 1.0 {
				t.Fatalf("bearing = %v, want ~%v", got, c.want)
			}
		})
	}
}

func TestInitialBearing_LondonToNY(t *testing.T) {
	// London → New York, ~288.3° per v1's spherical model.
	got := initialBearing(51.5074, -0.1278, 40.7128, -74.006)
	if math.Abs(got-288.3) > 1.0 {
		t.Fatalf("London→NY bearing = %v, want ~288.3", got)
	}
}

func TestHaversineKm(t *testing.T) {
	if d := haversineKm(51.5, -1, 51.5, -1); d != 0 {
		t.Fatalf("identical points = %v, want 0", d)
	}
	// London → Paris ~344 km (ceil rounding; ±2 km for endpoint precision).
	if d := haversineKm(51.5074, -0.1278, 48.8566, 2.3522); d < 342 || d > 346 {
		t.Fatalf("London→Paris = %v, want 342..346", d)
	}
}

func TestGridPath_InvalidGrids(t *testing.T) {
	for _, c := range []struct{ a, b string }{
		{"ZZ99", "JN58td"}, {"JN58td", "ZZ99"}, {"", "JN58td"}, {"JN58td", ""},
	} {
		if _, ok := GridPath(c.a, c.b); ok {
			t.Fatalf("GridPath(%q,%q) ok=true, want false", c.a, c.b)
		}
	}
}

func TestGridPath_ShortLongRelationship(t *testing.T) {
	geo, ok := GridPath("JN58td", "FN31pr")
	if !ok {
		t.Fatal("GridPath ok=false, want true")
	}
	// Short + long bearings are 180° apart.
	if d := math.Abs(math.Mod(geo.ShortBearingDeg+180, 360) - geo.LongBearingDeg); d > 0.2 {
		t.Fatalf("long bearing %v not opposite short %v", geo.LongBearingDeg, geo.ShortBearingDeg)
	}
	// Short + long distance ≈ Earth circumference.
	circ := math.Ceil(2 * math.Pi * earthRadiusKm)
	if sum := geo.ShortDistanceKm + geo.LongDistanceKm; math.Abs(sum-circ) > 2 {
		t.Fatalf("short+long = %v, want ~%v", sum, circ)
	}
	// Bearings rounded to 0.1°.
	if geo.ShortBearingDeg*10 != math.Round(geo.ShortBearingDeg*10) {
		t.Fatalf("short bearing %v not rounded to 0.1°", geo.ShortBearingDeg)
	}
}
